ledger/common/encoding/encoding.go

```diff
@@ -168,21 +184,36 @@ func EncodeKey(k *ledger.Key) []byte {
 
 // encodeKey encodes a key into a byte slice
 func encodeKey(k *ledger.Key) []byte {
-	buffer := make([]byte, 0)
+	buffer := make([]byte, 0, encodedKeyLength(k))
+	return encodeAndAppendKey(buffer, k)
+}
+
+func encodeAndAppendKey(buffer []byte, k *ledger.Key) []byte {
 	// encode number of key parts
 	buffer = utils.AppendUint16(buffer, uint16(len(k.KeyParts)))
+
 	// iterate over key parts
 	for _, kp := range k.KeyParts {
-		// encode the key part
-		encKP := encodeKeyPart(&kp)
 		// encode the len of the encoded key part
-		buffer = utils.AppendUint32(buffer, uint32(len(encKP)))
-		// append the encoded key part
-		buffer = append(buffer, encKP...)
+		buffer = utils.AppendUint32(buffer, uint32(encodedKeyPartLength(&kp)))
+
+		// encode the key part
+		buffer = encodeAndAppendKeyPart(buffer, &kp)
 	}
+
 	return buffer
 }
 
+func encodedKeyLength(k *ledger.Key) int {
+	// Key is encoded as: number of key parts (2 bytes) and for each key part,
+	// the key part size (4 bytes) + encoded key part (n bytes).
+	size := 2 + 4*len(k.KeyParts)
+	for _, kp := range k.KeyParts {
+		size += encodedKeyPartLength(&kp)
+	}
+	return size
+}
+
 // DecodeKey constructs a key from an encoded key part
 func DecodeKey(encodedKey []byte) (*ledger.Key, error) {
```

Comment:
do we need to add zeroCopy flag to the root method like this one and pass it to the underlying methods ?

Reply:
Not sure. I don't think `DecodeKey()` is currently used, but if we add zeroCopy flag to it, then:
- we need to add more tests for `DecodeKey()`
- and for consistency, also add zeroCopy flag to related functions, like `DecodeKeyPart()` and `DecodePayload()`, then add tests for them too

I'll add these to the TODO list of this PR.